PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/UseOfBlockOnAsyncClientsCheck.java

```diff
@@ -0,0 +1,163 @@
+
+                ANTI_PATTERN_MESSAGE = (String) RULE_CONFIGS_MAP.get(ANTI_PATTERN_MESSAGE_KEY);
+
+            } catch (IOException e) {
+                throw new RuntimeException("Error loading config file", e);
+            }
+        }
+
+        /**
+         * Constructor to initialize the visitor with the holder and isOnTheFly flag.
+         * @param holder ProblemsHolder
+         * @param isOnTheFly boolean
+         */
+        public UseOfBlockOnAsyncClientsVisitor(ProblemsHolder holder, boolean isOnTheFly) {
+            this.holder = holder;
+            this.isOnTheFly = isOnTheFly;
+        }
+
+        /**
+         * Visit method call expressions to check for the use of block() method on async clients.
+         * This method will check if the method call is block() and if it is used in an async context.
+         * @param expression
+         */
+        @Override
+        public void visitMethodCallExpression(@NotNull PsiMethodCallExpression expression) {
+            super.visitMethodCallExpression(expression);
+
+            // Check if the element is a method call expression
+            if (expression instanceof PsiMethodCallExpression) {
+                PsiMethodCallExpression methodCall = expression;
+
+                // Check if the method call is block() or a variant of block() method
+                if (methodCall.getMethodExpression().getReferenceName().startsWith("block")) {
+                    boolean isAsyncContext = checkIfAsyncContext(methodCall);
+
+                    // Check if the method call is on an Azure SDK client and is in an async context
+                    if (isAsyncContext && isAzureClient(methodCall)) {
+                        holder.registerProblem(expression, ANTI_PATTERN_MESSAGE);
+                    }
+                }
+            }
+        }
+
+        /**
+         * Helper method to check if the method call is within an async context.
+         * This method will check if the method call is on a reactive type like Mono, Flux, etc.
+         * @param methodCall
+         * @return true if the method call is on a reactive type, false otherwise
+         */
+        private boolean checkIfAsyncContext(@NotNull PsiMethodCallExpression methodCall) {
+            PsiExpression expression = methodCall.getMethodExpression().getQualifierExpression();
+
+            // Check if the method call is on a reactive type
+            if (expression != null) {
+                PsiType type = expression.getType();
+
+                // Check if the type is a reactive type
+                if (type != null) {
+                    String typeName = type.getCanonicalText();
+
+                    // Check for common async/reactive types directly in the list asyncReturnTypes
+                    for (String asyncReturnType : ASYNC_RETURN_TYPES) {
+                        if (typeName.startsWith(asyncReturnType)) {
```

Comment:
We should check if it's an instance of Mono or Flux because we may have custom types created like [CosmosPagedFlux](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/util/CosmosPagedFlux.java), [SearchPagedFlux](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/util/SearchPagedFlux.java) etc. So, `startsWith` generally doesn't capture all these scenarios.

Reply:
if these custom types have the "Flux" or "Mono" in the package name then a "package name contains Flux/Mono" should be enough?

Reply:
The package name may not contain Flux/Mono as CosmosPagedFlux is in `com.azure.cosmos.util` package and SearchPagedFlux is in `com.azure.search.documents.util` package.

Reply:
ooh so I have to check for a Flux/Mono return type -- this was my initial approach but I pivoted a few hours later because traversing up the tree was getting tricky but I can look into it again

Reply:
-- check if its an instance of Mono/Flux (rather than just the package name) -- use instanceof